### DISCUSSION TITLE
Error serializing pages Cannot read property 'nodes' of undefined

### DISCUSSION THREAD

**Discussion by MarcoAdinolfi** [gatsbyjs/gatsby#33345] [Help]
Hello,
I have a problem with the new version of gatsby-plugin-sitemap.
In my gatsby-config.js file I reported the following options that worked perfectly, generating the map in the siteUrl / sitemap.xml path:

```
module.exports = {
  siteMetadata: {
    title: `Dalyrium Bay`,
    description: `Benvenuto sul sito dei Dalyrium Bay! La nostra musica ti farà ballare`,
    siteUrl: `https://www.dalyriumbay.com`,
  },
  flags: { PRESERVE_WEBPACK_CACHE: true },
  plugins: [
    `gatsby-plugin-react-helmet`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `img`,
        path: `${__dirname}/src/img`,
      },
    },
    `gatsby-transformer-sharp`,
    `gatsby-plugin-sharp`,
    `gatsby-plugin-smoothscroll`,
    {
      resolve: `gatsby-plugin-sitemap`,
      options: {
        query: `
          {
            site {
              siteMetadata {
                title
                description
                author
                siteUrl
              }
            }

            allSitePage {
              nodes {
                path
              }
            }
        }`,
        serialize: ({ site, allSitePage }) =>
          allSitePage.nodes.map(node => {
            return {
              url: `${site.siteMetadata.siteUrl}${node.path}`,
              changefreq: `monthly`,
              priority: 0.7,
            }
          })
      }
    },
    `gatsby-plugin-sitemap`,
    {
      resolve: 'gatsby-plugin-web-font-loader',
      options: {
        google: {
          families: ['Roboto Mono:400', 'Nunito Sans:300,800', 'Nunito Sans:italic']
        }
      }
    },
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        name: `Dalyrium Bay`,
        short_name: `Dalyrium Bay`,
        lang: `it`,
        start_url: `/`,
        background_color: `#000000`,
        theme_color: `#000000`,
        display: `minimal-ui`,
        icon: `src/img/logo_Dalyrium.png`, // This path is relative to the root of the site.
      },
    },
    // this (optional) plugin enables Progressive Web App + Offline functionality
    // To learn more, visit: https://gatsby.dev/offline
    // `gatsby-plugin-offline`,
  ],
}

```

I upgraded to Gatsby v3 and updated dependencies and plugins, including gatsby-plugin-sitemap. Now trying to build I get the following error:
```
ERROR

[gatsby-plugin-sitemap]: Error serializing pages Cannot read property 'nodes' of undefined

  TypeError: Cannot read property 'nodes' of undefined
  
  - gatsby-config.js: 42 serialize
    /Users/marcoadinolfi/Documents/dalyriumbay-sito/gatsby-config.js:42:23
  
  - gatsby-node.js: 96 _callee $
    [dalyriumbay-site] / [gatsby-plugin-sitemap] /gatsby-node.js:96:36
  
  - task_queues.js: 93 processTicksAndRejections
    internal / process / task_queues.js: 93: 5
```

Why does this happen? I haven't touched anything in gatsby-config.js and as I said, it worked flawlessly before.

Furthermore, trying with the default settings, the sitemap is generated, but in the path 'siteUrl / sitemap / sitemap-0.xml' (so not in 'siteUrl/sitemap.xml' anymore), this is also something that I don't understand why it happens.

Can someone help me by answering the two questions and explaining what am I wrong?
Thank you very much

Further info:
Gatsby CLI version: 3.14.0
Gatsby version: 3.14.0
NPM version: 7.13.0
Node version: 14.16.0
"gatsby-plugin-sitemap": "^4.10.0"

**Comment by hgb123** [comment]
I came across this problem after upgrading to the latest packages too. After looking at the latest updates of the plugin, this is how I solved the problem, `resolvePages` first and `serialize` is now a callback function on each page

```
module.exports = {
  // ...
  plugins: [
    // ...
    {
      resolve: `gatsby-plugin-sitemap`,
      options: {
        query: `
          {
            site {
              siteMetadata {
                title
                description
                author
                siteUrl
              }
            }

            allSitePage {
              nodes {
                path
              }
            }
        }`,
        resolvePages: ({ site, allSitePage }) => {
          return allSitePage.nodes.map(node => ({
            url: `${site.siteMetadata.siteUrl}${node.path}`,
          }))
        },
        serialize: ({ url }) => {
          return {
            url,
            changefreq: `monthly`,
            priority: 0.7,
          }
        },
      },
    },
    // ...
  ],
}
```

**Reply by MarcoAdinolfi** [reply]
Hello,
Thank you so much for you answer.
I followed your advice and now my file is like this:
`module.exports = {
  siteMetadata: {
    title: `Dalyrium Bay`,
    description: `Benvenuto sul sito dei Dalyrium Bay! La nostra musica ti farà ballare`,
    siteUrl: `https://www.dalyriumbay.com`,
  },
  flags: { PRESERVE_WEBPACK_CACHE: true },
  plugins: [
    `gatsby-plugin-react-helmet`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `img`,
        path: `${__dirname}/src/img`,
      },
    },
    `gatsby-transformer-sharp`,
    `gatsby-plugin-sharp`,
    `gatsby-plugin-smoothscroll`,
    `gatsby-transformer-json`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        path: `./src/date-live/`,
      },
    },
    {
      resolve: `gatsby-plugin-sitemap`,
      options: {
        query: `
          {
            site {
              siteMetadata {
                title
                description
                author
                siteUrl
              }
            }

            allSitePage {
              nodes {
                path
              }
            }
        }`,
        resolvePages: ({ site, allSitePage }) => {
          return allSitePage.nodes.map(node => ({
            url: `${site.siteMetadata.siteUrl}${node.path}`,
          }))
        },
        serialize: ({ url }) => {
          return {
            url,
            changefreq: `monthly`,
            priority: 0.7,
          }
        },
      },
    },
    {
      resolve: 'gatsby-plugin-web-font-loader',
      options: {
        google: {
          families: ['Roboto Mono:400', 'Nunito Sans:300,800', 'Nunito Sans:italic']
        }
      }
    },
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        name: `Dalyrium Bay`,
        short_name: `Dalyrium Bay`,
        lang: `it`,
        start_url: `/`,
        background_color: `#000000`,
        theme_color: `#000000`,
        display: `minimal-ui`,
        icon: `src/img/logo_Dalyrium.png`, // This path is relative to the root of the site.
      },
    },
    // this (optional) plugin enables Progressive Web App + Offline functionality
    // To learn more, visit: https://gatsby.dev/offline
    // `gatsby-plugin-offline`,
  ],
}
`

but it still returns the error:
` ERROR #11321  PLUGIN

"gatsby-plugin-sitemap" threw an error while running the onPostBuild lifecycle:

[gatsby-plugin-sitemap]: Error in default page filter

  71 |
  72 |             reporter.verbose(_internals.REPORTER_PREFIX + " Filtering " + allPages.length + " pages based on " + excludes.length + "
 excludes");
> 73 |             _pageFilter = (0, _internals.pageFilter)({
     |                                                     ^
  74 |               allPages: allPages,
  75 |               filterPages: filterPages,
  76 |               excludes: excludes

File: node_modules/gatsby-plugin-sitemap/gatsby-node.js:73:53



  Error: [gatsby-plugin-sitemap]: Error in default page filter
  
  - internals.js:184 
    [dalyriumbay-sito]/[gatsby-plugin-sitemap]/internals.js:184:15
  
  - Array.some
  
  - internals.js:170 
    [dalyriumbay-sito]/[gatsby-plugin-sitemap]/internals.js:170:48
  
  - Array.filter
  
  - internals.js:169 pageFilter
    [dalyriumbay-sito]/[gatsby-plugin-sitemap]/internals.js:169:32
  
  - gatsby-node.js:73 _callee$
    [dalyriumbay-sito]/[gatsby-plugin-sitemap]/gatsby-node.js:73:53
  
  - task_queues.js:93 processTicksAndRejections
    internal/process/task_queues.js:93:5`

Do you understand what's wrong now?
Thank you again
